// astro/dist/compiler/transform/util/end-of-head.js
const validHeadElements = new Set(["!doctype", "title", "meta", "link", "style", "script", "noscript", "base"]);
class EndOfHead {
  constructor() {
    this.head = null;
    this.firstNonHead = null;
    this.parent = null;
    this.stack = [];
    this.append = () => void 0;
  }
  get found() {
    return !!(this.head || this.firstNonHead);
  }
  enter(node) {
    if (this.found) {
      return;
    }
    this.stack.push(node);
    if (!node.name) {
      return;
    }
    const name = node.name.toLowerCase();
    if (name === "head") {
      this.head = node;
      this.parent = this.stack[this.stack.length - 2];
      this.append = this.appendToHead;
      return;
    }
    if (!validHeadElements.has(name)) {
      this.firstNonHead = node;
      this.parent = this.stack[this.stack.length - 2];
      this.append = this.prependToFirstNonHead;
      return;
    }
  }
  leave(_node) {
    this.stack.pop();
  }
  appendToHead(...nodes) {
    var _a;
    const head = this.head;
    head.children = (_a = head.children) != null ? _a : [];
    head.children.push(...nodes);
  }
  prependToFirstNonHead(...nodes) {
    var _a, _b, _c;
    let idx = ((_a = this.parent) == null ? void 0 : _a.children.indexOf(this.firstNonHead)) || 0;
    (_c = (_b = this.parent) == null ? void 0 : _b.children) == null ? void 0 : _c.splice(idx, 0, ...nodes);
  }
}
export {
  EndOfHead
};
